caffe2/operators/collect_and_distribute_fpn_rpn_proposals_op.cc

```diff
@@ -0,0 +1,364 @@
+  auto lvls = utils::MapRoIsToFpnLevels(rois_block,
+                                        lvl_min, lvl_max,
+                                        canon_scale, canon_level);
+
+  // equivalent to python code
+  //   outputs[0].reshape(rois.shape)
+  //   outputs[0].data[...] = rois
+  auto* rois_out = Output(0);
+  rois_out->Resize(rois.rows(), rois.cols());
+  Eigen::Map<ERArrXXf> rois_out_mat(rois_out->mutable_data<float>(),
+                                    rois.rows(), rois.cols());
+  rois_out_mat = rois;
+
+  // Create new roi blobs for each FPN level
+  // (See: modeling.FPN.add_multilevel_roi_blobs which is similar but annoying
+  // to generalize to support this particular case.)
+  //
+  // equivalent to python code
+  //   rois_idx_order = np.empty((0, ))
+  //   for (output_idx, lvl in enumerate(range(lvl_min, lvl_max + 1)))
+  //       idx_lvl = np.where(lvls == lvl)[0]
+  //       blob_roi_level = rois[idx_lvl, :]
+  //       outputs[output_idx + 1].reshape(blob_roi_level.shape)
+  //       outputs[output_idx + 1].data[...] = blob_roi_level
+  //       rois_idx_order = np.concatenate((rois_idx_order, idx_lvl))
+  //   rois_idx_restore = np.argsort(rois_idx_order)
+  //   blob_utils.py_op_copy_blob(rois_idx_restore.astype(np.int32), outputs[-1])
+  EArrXi rois_idx_restore;
+  for (int i = 0, lvl = lvl_min; i < num_roi_lvls; i++, lvl++) {
+    ERArrXXf blob_roi_level;
+    EArrXi idx_lvl;
+    utils::RowsWhereRoILevelEquals(rois, lvls, lvl, &blob_roi_level, &idx_lvl);
+
+    // Output blob_roi_level
+    auto* roi_out = Output(i + 1);
+    roi_out->Resize(blob_roi_level.rows(), blob_roi_level.cols());
+    Eigen::Map<ERArrXXf> roi_out_mat(roi_out->mutable_data<float>(),
+                                     blob_roi_level.rows(),
+                                     blob_roi_level.cols());
+    roi_out_mat = blob_roi_level;
+
+    // Append indices from idx_lvl to rois_idx_restore
+    rois_idx_restore.conservativeResize(rois_idx_restore.size() + idx_lvl.size());
+    rois_idx_restore.tail(idx_lvl.size()) = idx_lvl;
+  }
+  utils::ArgSort(rois_idx_restore);
+  auto* rois_idx_restore_out = Output(OutputSize() - 1);
+  rois_idx_restore_out->Resize(rois_idx_restore.size());
+  Eigen::Map<EArrXi> rois_idx_restore_out_mat(rois_idx_restore_out->mutable_data<int>(),
+                                              rois_idx_restore.size());
+  rois_idx_restore_out_mat = rois_idx_restore;
+
+  return true;
+}
+
+namespace {
+
+REGISTER_CPU_OPERATOR(CollectAndDistributeFpnRpnProposals, CollectAndDistributeFpnRpnProposalsOp<CPUContext>);
+
+#ifdef CAFFE2_HAS_MKL_DNN
+REGISTER_MKL_OPERATOR(
+    CollectAndDistributeFpnRpnProposals,
+    mkl::MKLFallbackOp<CollectAndDistributeFpnRpnProposalsOp<CPUContext>>);
+#endif // CAFFE2_HAS_MKL_DNN
+
+OPERATOR_SCHEMA(CollectAndDistributeFpnRpnProposals)
+    .NumInputs(2, INT_MAX)
+    .NumOutputs(3, INT_MAX)
+    .SetDoc(R"DOC(
+Merge RPN proposals generated at multiple FPN levels and then
```

Comment:
Maybe add two notes to the op documentation:
1) Only inference is supported (the Python version has a training mode in which it dynamically builds a training minibatch for the detection head(s) based on the FPN RPN proposals)
2) The inputs documented below are only an example; if the min/max levels are changed, then the number of inputs and outputs, as well as their level numbering, will change